File: backend/src/main/java/ma/investpro/entity/DepenseInvestissement.java
package ma.investpro.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * Entité représentant une dépense d'investissement
 */
@Entity
@Table(name = "depenses_investissement", indexes = {
    @Index(name = "idx_depense_numero_facture", columnList = "numeroFacture"),
    @Index(name = "idx_depense_date_facture", columnList = "dateFacture"),
    @Index(name = "idx_depense_fournisseur", columnList = "fournisseur_id"),
    @Index(name = "idx_depense_projet", columnList = "projet_id")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class DepenseInvestissement extends BaseEntity {

    @Column(nullable = false, length = 100)
    @NotBlank(message = "Le numéro de facture est obligatoire")
    @Size(max = 100, message = "Le numéro de facture ne peut pas dépasser 100 caractères")
    private String numeroFacture;

    @Column(nullable = false)
    @NotNull(message = "La date de facture est obligatoire")
    private LocalDate dateFacture;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "fournisseur_id", nullable = false)
    @NotNull(message = "Le fournisseur est obligatoire")
    private Fournisseur fournisseur;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "projet_id", nullable = false)
    @NotNull(message = "Le projet est obligatoire")
    private Projet projet;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "axe_analytique_id")
    private AxeAnalytique axeAnalytique;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "convention_id")
    private Convention convention;

    // Montants de la facture
    @Column(nullable = false, precision = 15, scale = 2)
    @NotNull(message = "Le montant HT est obligatoire")
    @DecimalMin(value = "0.00", message = "Le montant HT doit être positif")
    private BigDecimal montantHt;

    @Column(nullable = false, precision = 5, scale = 2)
    @NotNull(message = "Le taux de TVA est obligatoire")
    @DecimalMin(value = "0.00", message = "Le taux de TVA doit être positif")
    private BigDecimal tauxTva = new BigDecimal("20.00");

    @Column(nullable = false, precision = 15, scale = 2)
    @NotNull(message = "Le montant TVA est obligatoire")
    @DecimalMin(value = "0.00", message = "Le montant TVA doit être positif")
    private BigDecimal montantTva;

    @Column(nullable = false, precision = 15, scale = 2)
    @NotNull(message = "Le montant TTC est obligatoire")
    @DecimalMin(value = "0.00", message = "Le montant TTC doit être positif")
    private BigDecimal montantTtc;

    // Références marché / décompte
    @Column(length = 100)
    @Size(max = 100, message = "La référence marché ne peut pas dépasser 100 caractères")
    private String referenceMarche;

    @Column(length = 100)
    @Size(max = 100, message = "Le numéro de décompte ne peut pas dépasser 100 caractères")
    private String numeroDecompte;

    // Retenues
    @Column(precision = 15, scale = 2)
    @DecimalMin(value = "0.00", message = "La retenue TVA doit être positive")
    private BigDecimal retenueTva = BigDecimal.ZERO;

    @Column(precision = 15, scale = 2)
    @DecimalMin(value = "0.00", message = "La retenue IS tiers doit être positive")
    private BigDecimal retenueIsTiers = BigDecimal.ZERO;

    @Column(precision = 15, scale = 2)
    @DecimalMin(value = "0.00", message = "La retenue non-résident doit être positive")
    private BigDecimal retenueNonResident = BigDecimal.ZERO;

    @Column(precision = 15, scale = 2)
    @DecimalMin(value = "0.00", message = "La retenue garantie doit être positive")
    private BigDecimal retenueGarantie = BigDecimal.ZERO;

    // Informations de paiement
    @Column
    private LocalDate datePaiement;

    @Column(length = 100)
    @Size(max = 100, message = "La référence de paiement ne peut pas dépasser 100 caractères")
    private String referencePaiement;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "compte_bancaire_id")
    private CompteBancaire compteBancaire;

    @Column(nullable = false)
    private Boolean paye = false;

    @Column(columnDefinition = "TEXT")
    private String remarques;

    // Calculs automatiques
    @PrePersist
    @PreUpdate
    private void calculateAmounts() {
        // Calcul TVA si pas déjà défini
        if (montantHt != null && tauxTva != null) {
            montantTva = montantHt.multiply(tauxTva).divide(new BigDecimal("100"), 2, BigDecimal.ROUND_HALF_UP);
            montantTtc = montantHt.add(montantTva);
        }

        // Calcul automatique retenue garantie 10% si non défini
        if (montantHt != null && retenueGarantie == null) {
            retenueGarantie = montantHt.multiply(new BigDecimal("0.10")).setScale(2, BigDecimal.ROUND_HALF_UP);
        }

        // Calcul IS tiers 10% pour fournisseurs non-résidents
        if (fournisseur != null && fournisseur.getNonResident() && montantHt != null) {
            retenueIsTiers = montantHt.multiply(new BigDecimal("0.10")).setScale(2, BigDecimal.ROUND_HALF_UP);
        }
    }

    /**
     * Calcule le montant net à payer
     */
    public BigDecimal getMontantNetAPayer() {
        BigDecimal montantNet = montantTtc;
        if (retenueTva != null) montantNet = montantNet.subtract(retenueTva);
        if (retenueIsTiers != null) montantNet = montantNet.subtract(retenueIsTiers);
        if (retenueNonResident != null) montantNet = montantNet.subtract(retenueNonResident);
        if (retenueGarantie != null) montantNet = montantNet.subtract(retenueGarantie);
        return montantNet;
    }
}

File: backend/src/main/kotlin/ma/investpro/entity/DepenseInvestissement.kt
package ma.investpro.entity

import jakarta.persistence.*
import jakarta.validation.constraints.*
import java.math.BigDecimal
import java.time.LocalDate

@Entity
@Table(
    name = "depenses_investissement",
    indexes = [
        Index(name = "idx_depenses_numero_facture", columnList = "numero_facture"),
        Index(name = "idx_depenses_date_facture", columnList = "date_facture"),
        Index(name = "idx_depenses_fournisseur", columnList = "fournisseur_id"),
        Index(name = "idx_depenses_projet", columnList = "projet_id"),
        Index(name = "idx_depenses_axe", columnList = "axe_analytique_id"),
        Index(name = "idx_depenses_convention", columnList = "convention_id"),
        Index(name = "idx_depenses_paye", columnList = "paye")
    ]
)
class DepenseInvestissement(
    @Column(name = "numero_facture", nullable = false, length = 100)
    @field:NotBlank
    var numeroFacture: String = "",

    @Column(name = "date_facture", nullable = false)
    @field:NotNull
    var dateFacture: LocalDate = LocalDate.now(),

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fournisseur_id", nullable = false)
    var fournisseur: Fournisseur? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "projet_id", nullable = false)
    var projet: Projet? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "axe_analytique_id")
    var axeAnalytique: AxeAnalytique? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "convention_id")
    var convention: Convention? = null,

    // Montants facture
    @Column(name = "montant_ht", nullable = false, precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var montantHt: BigDecimal = BigDecimal.ZERO,

    @Column(name = "taux_tva", nullable = false, precision = 5, scale = 2)
    @field:DecimalMin("0.00")
    var tauxTva: BigDecimal = BigDecimal("20.00"),

    @Column(name = "montant_tva", nullable = false, precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var montantTva: BigDecimal = BigDecimal.ZERO,

    @Column(name = "montant_ttc", nullable = false, precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var montantTtc: BigDecimal = BigDecimal.ZERO,

    // Références
    @Column(name = "reference_marche", length = 100)
    var referenceMarche: String? = null,

    @Column(name = "numero_decompte", length = 100)
    var numeroDecompte: String? = null,

    // Retenues
    @Column(name = "retenue_tva", precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var retenueTva: BigDecimal = BigDecimal.ZERO,

    @Column(name = "retenue_is_tiers", precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var retenueIsTiers: BigDecimal = BigDecimal.ZERO,

    @Column(name = "retenue_non_resident", precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var retenueNonResident: BigDecimal = BigDecimal.ZERO,

    @Column(name = "retenue_garantie", precision = 15, scale = 2)
    @field:DecimalMin("0.00")
    var retenueGarantie: BigDecimal = BigDecimal.ZERO,

    // Paiement
    @Column(name = "date_paiement")
    var datePaiement: LocalDate? = null,

    @Column(name = "reference_paiement", length = 100)
    var referencePaiement: String? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "compte_bancaire_id")
    var compteBancaire: CompteBancaire? = null,

    @Column(nullable = false)
    var paye: Boolean = false,

    @Column(columnDefinition = "TEXT")
    var remarques: String? = null,

    // Champs de gestion avancée
    @Column(name = "type_depense", length = 20)
    @Enumerated(EnumType.STRING)
    var typeDepense: TypeDepense = TypeDepense.STANDARD,

    @Column(name = "statut", length = 20)
    @Enumerated(EnumType.STRING)
    var statut: StatutDepense = StatutDepense.EN_COURS,

    @Column(name = "taux_commission", precision = 5, scale = 2)
    @field:DecimalMin("0.00")
    @field:DecimalMax("100.00")
    var tauxCommission: BigDecimal? = null,

    @Column(name = "base_calcul", length = 10)
    @Enumerated(EnumType.STRING)
    var baseCalcul: BaseCalcul = BaseCalcul.TTC,

    @Column(name = "objet", columnDefinition = "TEXT")
    var objet: String? = null,

    @Column(name = "date_demarrage")
    var dateDemarrage: LocalDate? = null,

    @Column(name = "delai_mois")
    var delaiMois: Int? = null,

    @Column(name = "date_fin_prevue")
    var dateFinPrevue: LocalDate? = null,

    @Column(name = "designation", length = 500)
    var designation: String? = null

) : BaseEntity()

enum class TypeDepense {
    STANDARD,
    CADRE,
    NON_CADRE,
    SPECIFIQUE,
    AVENANT
}

enum class StatutDepense {
    VALIDEE,
    EN_COURS,
    ACHEVE,
    EN_RETARD,
    ANNULE
}

enum class BaseCalcul {
    TTC,
    HT
}
